### PR TITLE
Use generation blocks for reliable SFT labels

### LINKED ISSUE #1804: Thinker chat templates cannot be used for SFT: every row's labels are undecidable
## Summary

The `*_thinker*` chat templates rewrite assistant content while rendering, so no span-based
derivation can locate the assistant turn. After the detection added for #1800, this no longer
corrupts labels silently — but it means **every conversation using these templates is dropped**,
so the templates are effectively unusable for SFT.

Affected (verified against the bundled test tokenizer, both `eos_token` variants, both
alternating and consecutive-assistant conversation shapes — 16 of 44 swept combinations):

- `olmo_thinker`
- `olmo_thinker_remove_intermediate_thinking`
- `tulu_thinker`
- `tulu_thinker_r1_style`

## Why derivation fails

Both available derivations locate an assistant turn by re-rendering conversation prefixes —
either as strings (char offsets into the full render) or as token counts. Both assume the turn
appears in the full render the same way it appears in a prefix render. The thinker templates
break that assumption directly: they split on `</think>`, strip intermediate thinking, or inject
a `<think>` opener into the generation prompt, so the rendered turn is not the message content.

Before the #1800 fix this was **silent**. The token-count path produced spans that started
mid-content:

```
olmo_thinker, alternating:
  trained='<|im_end|>WO<|endoftext|>'
  # ASSISTONE dropped entirely; only the fragment "WO" of ASSISTTWO trained

olmo_thinker, thinking content:
  trained='THINKONE</think>ANSWERONE<|im_end|>THINKTWO</think>ANSWERTWO<|endoftext|>'
  # the opening <think> tag masked out of the loss
```

That second case is the one worth keeping in view: a thinking model that never sees `<think>` in
its loss. It is now detected rather than trained on, but it is not *fixed*.

## Related: consecutive assistant turns

The same root cause makes consecutive assistant turns undecidable for **every** template
including `tulu` and `olmo`, because deriving the second turn's boundary requires rendering a
prefix that ends in an assistant turn — which templates special-casing the final turn render
differently from the full conversation. Rare in practice (15 rows of 939k in
`tulu-3-sft-olmo-2-mixture`) and currently dropped, but a `{% generation %}` fix would resolve
it too.

## Proposed fix

Mark assistant content in each template with `{% generation %}` blocks and read the mask from
`apply_chat_template(..., return_assistant_tokens_mask=True)`. This removes the offset
arithmetic entirely: no prefix re-rendering, no prefix-stability requirement, and content
rewriting stops mattering because the template itself declares which span is trainable.

Verified working on the pinned transformers (5.4.0): the returned `assistant_masks` aligns 1:1
with `input_ids` and survives `truncation=True, max_length=N`.

### Two constraints any patch must respect

1. **A template with no `{% generation %}` block does not raise.** It warns and returns an
   all-zero mask. A partial migration would therefore train on nothing, silently — strictly
   worse than today's behaviour, where undecidable rows are detected and counted. Any patch
   needs an explicit non-empty-mask assertion. There is already a regression test pinning this
   behaviour (`test_generation_blocks_yield_an_all_zero_mask_without_raising`).
2. **External templates have no generation blocks.** Qwen3's and Olmo-3's published templates
   are used whenever `--chat_template_name` is unset or unrecognised, so the existing derivation
   cannot simply be deleted; it has to stay as the path for templates the repo does not own.

## Test scaffolding already in place

`TestChatTemplateAssistantLabelSweep` in `open_instruct/test_dataset_transformation.py` sweeps
11 SFT templates × 2 `eos_token` variants × 2 conversation shapes, fully offline against the
bundled tokenizer fixture. `DERIVABLE_COMBINATIONS` lists the 18 of 44 that currently produce
correct labels; the other 26 are asserted to train nothing rather than train the wrong tokens.

Fixing this issue means moving entries into `DERIVABLE_COMBINATIONS` and watching the sweep go
green — the specification is already encoded, so the work is verifiable as it lands.

## Priority

Not urgent. These templates are currently unusable rather than dangerous, and no shipped Olmo 3
model was affected (the published SFT datasets were verified correctly masked — see the
resolution comment on #1800). But `olmo_thinker` is the natural template for thinking SFT, so
this blocks that path.

Refs #1800.

### PR DESCRIPTION
**Pull request by farhatkevin**
## Summary
- mark assistant output in repository-owned SFT templates with `{% generation %}` blocks and use Transformers' aligned assistant mask, fixing thinker templates that rewrite content while rendering
- retain the verified prefix-derivation fallback for external tokenizer templates without generation blocks, and reject declared blocks that unexpectedly return an empty mask
- make consecutive assistant turns and truncation-safe last-turn labeling explicit in the existing template sweep

Fixes #1804.

## Test plan
- [x] `uv run pytest open_instruct/test_dataset_transformation.py -q --tb=short` (91 passed)
- [x] `make style && make quality`
- [x] verify all 11 SFT templates render byte-for-byte identically to `main`, with and without `add_generation_prompt`

GPU_TESTS=bypass

Made with [Cursor](https://cursor.com)

### CODE REVIEW

**Comment by farhatkevin** [comment]
@codex review

File: CHANGELOG.md [hunk 1]
```diff
@@ -10,6 +10,7 @@ All notable changes to this project will be documented in this file.
 - Drop stale async rollout results whose generating policy is more than `async_steps` behind the trainer (`max_result_age_steps`), replenishing a fresh prompt and logging a `stale_results_dropped` metric (https://github.com/allenai/open-instruct/pull/1738).
 
 ### Fixed
+- SFT can now label thinker chat templates (`olmo_thinker`, `tulu_thinker`, and the other in-repo SFT templates): assistant content is marked with `{% generation %}` and read via `return_assistant_tokens_mask=True`, so prefix arithmetic is no longer required. Templates without generation blocks still use the existing derivation, and an empty mask on a template that declared blocks is treated as underivable rather than silently training on nothing (https://github.com/allenai/open-instruct/issues/1804).
```
[the remark this answers is not in the document]
Comment (farhatkevin) [reply CHANGELOG.md]: Already addressed in 04afa39bc: the changelog entry now links to https://github.com/allenai/open-instruct/pull/1879.

File: open_instruct/dataset_transformation.py [hunk 2]
```diff
@@ -1330,6 +1375,64 @@ def _tokenize_tulu_sft_with_assistant_labels(
             f"`return_offsets_mapping` to derive assistant label spans, but got a slow tokenizer "
             f"({type(tokenizer).__name__}). Load the tokenizer with `use_fast=True`."
         )
+    # Templates that declare `{% generation %}` blocks give a 1:1 assistant mask.
+    # Hugging Face returns an all-zero mask (no raise) when the tags are missing, so
+    # only take this path when the template actually has them; otherwise fall through
+    # to prefix derivation, which is what published Qwen/Olmo templates still need.
+    template = tokenizer.chat_template
+    if isinstance(template, dict):
+        template = template.get("default") or ""
```
[the remark this answers is not in the document]
Comment (farhatkevin) [reply open_instruct/dataset_transformation.py]: Addressed in 385059298. The generation-mask path now calls `tokenizer.get_chat_template(tools=tools)`, matching Transformers’ selection of `tool_use` versus `default`. Added a regression test with a template dictionary whose generation blocks exist only in `tool_use`; 92 dataset-transformation tests pass.

**Comment by farhatkevin** [comment]
@codex review

File: open_instruct/dataset_transformation.py [hunk 3]
```diff
@@ -1330,6 +1375,62 @@ def _tokenize_tulu_sft_with_assistant_labels(
...
+                    "returned an all-zero assistant mask, so labels would train on nothing."
+                )
+            labels = torch.full_like(input_ids, MASKED_TOKEN_VALUE)
+            return input_ids, attention_mask, labels
+        if last_turn_only:
+            ones = assistant_masks[0].nonzero(as_tuple=True)[0]
+            end = int(ones[-1]) + 1
+            start = end - 1
+            while start > 0 and assistant_masks[0, start - 1]:
+                start -= 1
```
[the remark this answers is not in the document]
Comment (farhatkevin) [reply open_instruct/dataset_transformation.py]: Fixed in 66b0c3ede. Last-turn labeling no longer infers a turn by scanning a flattened mask: registered templates use exact generation ranges, while arbitrary external templates use the verified prefix derivation when generation-block ownership is ambiguous. Added regressions for contiguous blocks, multiple blocks per turn, earlier-block leakage, and truncation.

**Comment by farhatkevin** [comment]
@codex review

Additional adversarial audit fixes are in 66b0c3ede: exact generation ranges handle left truncation and zero-width blocks, external last-turn templates fall back safely when ownership is ambiguous, no-assistant rows remain fully masked, and structured `simple_chat` content is rejected rather than stringified.

**Review by abhishekraok — COMMENTED** [review]
Claude Code · session 6b836cd5 (claude.ai/code), posted on behalf of @abhishekraok

*Edited: shortened, and corrected one item I got wrong. My original review was written against a pre-rebase snapshot, so its claim about a #1876 collision was stale — `ded1913d0` already handles it. Verified: `terminate` on an over-length row through the generation path now writes a trainable EOS correctly. Sorry for the noise.*

The template rewrite holds up. All 11 templates render byte-identically to `main` across seven conversation shapes, both `eos` variants, `add_generation_prompt` either way, with and without tools. The line 1382 comment justifying the bypass of Transformers' flattened mask is accurate against 5.4.0. Two blocking items, both reproduced on `ded1913d0` with the bundled fixture tokenizer.

**1. `--add_bos` sends DPO back to the broken path.** `is_registered_template` (1385) is an exact string match, and `tc.add_bos` prepends `{{ bos_token }}` (749, 817, 895), so the gate fails and `last_turn_only` falls through to prefix derivation. For OLMo tokenizers with a non-`olmo*` template name `add_bos` is *asserted required* (838), so this is the production DPO config.

```
tulu_thinker  add_bos=False -> '<think>THINKTWO</think>ANSWERTWO<|end_of_text|>'
tulu_thinker  add_bos=True  -> 'THINKTWO</think>ANSWERTWO<|end_of_text|>'
```

`<think>` is dropped from the labels and the row is **kept**, so it mislabels rather than drops. The fix is already in your code: the `len(ranges) == assistant_turn_count` check (1419) is the invariant that makes `ranges[-1]` trustworthy, and it holds for all 11 templates with or without a bos prefix. Gate on that and drop the registry lookup. Nothing in the sweep sets `add_bos`.

**2. A block that excludes the closing token now trains no EOS.** For `last_turn_only=False` the new path trusts the block verbatim. A hub template shaped `...assistant\n{% generation %}{{ content }}{% endgeneration %}<|im_end|>\n` used to train through the closing token; now:

```
-> '<think>THINKONE</think>ANSWERONE<think>THINKTWO</think>ANSWERTWO'
```

No terminator in the loss, no warning, because the mask is non-empty. Same never-stop failure as #1876 from another direction. Suggest raising or logging when a range does not end on EOS or a special token.

**Worth noting, not blocking**

- `{% generation %}` now ships in released `tokenizer_config.json` / `chat_template.jinja` via the save paths and `export_chat_template.py`. transformers and vLLM 0.19.1 handle it; plain `jinja2` raises `unknown tag 'generation'`, and TGI's strip is spelling-sensitive. Either strip on save or state the cost.
- `DATASET_CACHE_VERSION` stays `v7`. In-repo templates get fresh caches since their strings changed, but a hub template already carrying generation blocks hashes identically to `main` while labelling differently.
- `docs/olmo3.md` still describes this bug as open. Also see #1882: the hub `think-dev` tokenizer is untouched by this PR and still breaks.

**Cleanups**

- Two labelling loops, already divergent: the prefix one skips zero-width offsets (1514), the new one (1426-1429) does not. Fast tokenizers emit those. Extract one helper.
- Three copies of the tokenize call with the same kwargs; the mask-out copy omits `return_offsets_mapping`.
- The new loop is O(ranges × tokens) with scalar tensor writes: 456 ms on a 65536-token row with 50 ranges, vs 29 ms for a numpy boolean mask with the identical predicate. Also `.tolist()` on offsets costs 10.4 ms/64K row vs 0.08 ms for `.numpy()`.
- `assistant_turn_count` (1410) re-derives `_trainable_assistant_indices` with `.get('role')` instead of `['role']`.
- The raise at 1423 is unreachable as gated; fixing item 1 makes it meaningful.
- In `tulu_thinker` and `tulu_thinker_r1_style` both branches of `{% if not loop.last %}` are now identical.
- Nothing pins the two invariants the PR rests on: tags are output-neutral, and one range per assistant turn. A parametrized test over templates × shapes would make the byte-for-byte claim CI.
- `test_one_bad_row_does_not_abort_a_dataset_map` now uses a separate tokenizer for the bad row, so no registered-template row can raise. `test_generation_blocks_yield_an_all_zero_mask_without_raising` uses a template with no generation blocks and passes for any implementation.

**Scope.** RL is being deprecated here in favour of miles, so only SFT and DPO matter, and of the scripts passing an in-repo thinker template exactly one is SFT/DPO (`scripts/train/olmo3/7b_think_dpo.sh`, the historical Olmo 3 recipe). Worth deciding whether these templates are legacy before investing in the above. #1882 lays out the options; your call.

**Comment by farhatkevin** [comment]
Thanks for the detailed review. I reproduced and fixed both blockers.

BOS-prefixed repository templates now retain generation-range labeling for DPO. External templates use stable renders before and through the final assistant message to establish block ownership. I didn’t use block-count equality alone: two blocks can belong to the first assistant while the final assistant emits none.

EOS or `<|im_end|>` immediately outside a selected generation block now produces an actionable error, with the row logged and dropped. Separate terminator blocks remain supported, and truncation still follows `over_length_strategy`.

I also bumped the cache version, consolidated tokenization and vectorized masking, fixed zero-width handling, removed duplicate branches, and strengthened the tests. All 11 templates are now checked against their pre-PR rendered output.

On scope, this is relevant to Olmo 3.5 too. Its template assembles reasoning, answers, and tool calls; the generation block should include that complete assistant output and its terminator—`<|im_end|>` for tool handoffs, EOS for answers. The inference-only `<think>` prompt stays outside. Its stable turn endings allow reliable last-turn labeling.

Added an annotated Olmo 3.5 fixture and validated it with the actual tokenizer. **Publishing those annotations and syncing the Hub tokenizer config remain separate work.** This PR doesn’t fix unannotated Hub templates, so #1882 remains relevant. The docs now explain that limitation and the generation-extension requirement for exported templates.

Validation: 149 dataset tests passed; formatting, linting, type checks, and docs build passed. The annotated Olmo 3.5 template matched the published template across 112 render checks.

**Comment by farhatkevin** [comment]
Follow-up: the Olmo 3.5 generation annotations are now published directly to `main` in [allenai/dolma2-tokenizer-olmo35, commit 8b971706](https://huggingface.co/allenai/dolma2-tokenizer-olmo35/commit/8b9717061fae09d5be814373d189919a62a9a00d).

The source `chat_template.jinja` and `tokenizer_config.json` are synced. Each assistant generation block includes reasoning, answer/tool-call output, and the appropriate terminator; assistant headers, separator newlines, and the inference-only generation prompt remain outside.

Validation: all 26 tokenizer regression tests and the six smoke checks passed, the SFT/DPO label tests passed with the actual tokenizer, and 112 before/after renders and token sequences were identical. Published files were read back and verified against the tested files.

This completes the Olmo 3.5 Hub follow-up mentioned in my previous comment. Training still needs this PR's labeling changes to consume the annotations correctly; #1882 remains relevant to other unannotated Hub templates. The PR docs now link to the published tokenizer commit (a769f68af).

**Review by abhishekraok — APPROVED** [review]
Claude Code · session for @abhishekraok

Approving. Verified locally on the PR head (149 tests pass with the LFS fixtures; labels correct on the Olmo 3.5 fixture for single-think, multi-turn with last-turn-only reasoning, `reasoning_content`, tool-call hand-off, and consecutive assistant turns, for SFT and DPO; instruct-dev without blocks still matches main; renders byte-identical for all 11 templates). For the record: this PR **trains** the opening `<think>`; think-template-label-spans (#1885), which masked it, is dropped in favour of this.

**Please fix before merge**
- `_labels_from_char_spans` indexes `offsets[:, 0]` unconditionally. A row that renders to zero tokens (user-only conversation under `assistant_message_only`) gives a 1-D empty array and raises `IndexError`, which is not `AssistantSpanDerivationError`, so it escapes the per-row catch and aborts `dataset.map`. One-line empty guard, as `_was_truncated` does.

**Requested, here or as an immediate follow-up**
- The block path runs none of main's span checks (starts after header, covers content, stays in turn). Reproduced: a block wrapping only the assistant header trains `<|assistant|>\n` and nothing else, silently. Our templates are placed correctly; external Hub templates are where this will bite. Port the header-bound and coverage checks to the char-range path.
- `_was_truncated` only detects a dropped tail, so with `truncation_side="left"` the `drop`/`terminate` strategies never fire, while the PR now adds left-truncation tests. Pre-existing, but the claim is new.
- Two small ones: one CHANGELOG sentence naming zephyr (the newline after eos is no longer trained; the only label change not from the #1804 fix), and a warning when a special token other than eos/`<|im_end|>` immediately follows a block.

**Body/doc nits**
- State the minimum Hub revision (8b971706); older olmo35 revisions silently fall back to the prefix path and mask the tag. Link hub-think-dev-breaks-labels (#1882) as intentionally still open. Note that `render_jinja_template` is a transformers-internal API.
- Not from this PR, but relevant to anyone reasoning about token boundaries here: tokenizer-loader-regression (#1896).

**Comment by farhatkevin** [comment]
Thanks @abhishekraok — pushed the fixes in 61528f6ee.

Fixed the empty-render crash and added a regression test showing that a user-only row under `assistant_message_only` gets filtered without aborting `dataset.map`.

Also fixed left-truncation detection: `drop` now removes those rows. `terminate` preserves the intact tail when truncating from the left, so it doesn't overwrite a tool-handoff token with EOS. Bumped the dataset cache version for the changed behavior.

Added the warning for other special tokens just outside generation blocks, the Zephyr newline changelog note, and the docs clarifications: Olmo 3.5 needs tokenizer revision `8b971706` or later with the annotations; #1882 intentionally remains open; `render_jinja_template` is a Transformers-internal API. Also linked #1896.

I'm leaving the broader header/content/turn-boundary validation for external templates as the follow-up you allowed. It needs to handle the Olmo 3.5 template assembling reasoning and tool calls, rather than assuming the rendered output equals the message's `content`. That validation is still outstanding and is now called out in the docs. The opening `<think>` remains trained.

Validation: all 156 dataset-transformation tests passed across runs. The five cache tests were run with in-process map/filter calls and a temporary cache because the local sandbox blocks multiprocessing sockets. Formatting, lint, type checks, and the docs build passed.
